src/main/java/net/neoforged/neoforge/client/renderstate/RegisterRenderStateModifiersEvent.java

```diff
@@ -0,0 +1,112 @@
+        ensureParametersMatchBounds(baseRenderer);
+        RenderStateExtensions.registerEntity(baseRenderer.getRawType(), modifier);
+    }
+
+    /**
+     * Registers a render state modifier for {@link MapRenderState}s which are run after the texture has been set
+     * and before decorations have been added. Can add custom data to the map using
+     * {@link IRenderStateExtension#setRenderData(ContextKey, Object)}.
+     *
+     * @param modifier The function for modifying the {@link net.minecraft.client.renderer.state.MapRenderState} and adding custom render data.
+     */
+    public void registerMapModifier(BiConsumer<MapItemSavedData, MapRenderState> modifier) {
+        RenderStateExtensions.registerMap(modifier);
+    }
+
+    /**
+     * Registers a render state modifier for {@link MapRenderState.MapDecorationRenderState}s which are run after
+     * vanilla map decoration data has been set. Can add custom data to the map using
+     * {@link IRenderStateExtension#setRenderData(ContextKey, Object)}.
+     *
+     * @param mapDecorationTypeKey Key for the registered {@link MapDecorationType}
+     * @param modifier             The function for modifying the {@link MapRenderState.MapDecorationRenderState} and adding custom render data.
+     */
+    public void registerMapDecorationModifier(ResourceKey<MapDecorationType> mapDecorationTypeKey, MapDecorationRenderStateModifier modifier) {
+        RenderStateExtensions.registerMapDecoration(mapDecorationTypeKey, modifier);
+    }
+
+    private static void ensureParametersMatchBounds(TypeToken<?> baseRenderer) {
+        if (baseRenderer.getType() instanceof ParameterizedType parameterizedType) {
+            Class<?> bound = baseRenderer.getRawType();
+            var parameterized = parameterizedType;
+            do {
+                var userArgs = parameterizedType.getActualTypeArguments();
+                var typeArgs = ((Class) parameterizedType.getRawType()).getTypeParameters();
+                for (int i = 0; i < userArgs.length; i++) {
+                    var userType = userArgs[i];
+                    // ignore wildcards. compilation would catch illegal states, so this must not be an important type to capture
+                    if (userType instanceof WildcardType) continue;
+                    var typeBounds = typeArgs[i].getBounds();
+                    // a generic class with multiple bounds, ie AbstractHoglinRenderer<T extends Mob & HoglinBase>
+                    if (userType instanceof TypeVariable<?> typeVariable) {
+                        var userBounds = typeVariable.getBounds();
+                        // check interface bounds
+                        for (int j = 0; j < typeBounds.length; j++) {
+                            if (!((Class) typeBounds[j]).isAssignableFrom((Class<?>) userBounds[j])) {
```

Comment:
Why are these Class casts safe? I'm pretty sure nothing guarantees that those won't be some other Type.